traitsui/editor.py

```diff
@@ -445,146 +512,121 @@ def sync_value(
-                        if key not in self._no_trait_update:
-                            self._no_trait_update[key] = None
-                            n = event.index
-                            try:
-                                eval(user_value)[
-                                    n : n + len(event.removed)
-                                ] = event.added
-                            except:
-                                from traitsui.api import raise_to_debug
-
-                                raise_to_debug()
-                            del self._no_trait_update[key]
-
-                    self.on_trait_change(
-                        editor_list_modified, editor_name + "_items"
-                    )
-                    self._user_from.append(
-                        (editor_name + "_items", editor_list_modified)
-                    )
-
-                if mode == "to" and not is_event:
-                    try:
-                        setattr(
-                            eval(user_ref),
-                            user_name,
-                            getattr(self, editor_name),
-                        )
-                    except:
-                        from traitsui.api import raise_to_debug
-
-                        raise_to_debug()
+        if is_list:
+
+            def user_list_modified(event):
+                if (
+                    isinstance(event, TraitListEvent)
+                    and key not in self._no_trait_update
+                ):
+                    with self.no_trait_update(key), self.raise_to_debug():
+                        n = event.index
+                        getattr(self, editor_name)[
+                            n : n + len(event.removed)
+                        ] = event.added
+
+            items = xuser_name + "_items"
+            user_object.on_trait_change(user_list_modified, items)
+            self._user_to.append((user_object, items, user_list_modified))
+
+    def _bind_to(self, key, user_object, xuser_name, editor_name, is_list):
+        """ Bind trait change handlers from a user object to the editor.
+
+        Parameters
+        ----------
+        key : str
+            The key to use to guard against recursive updates.
+        user_object : object
+            The extended name of the trait to be used on the user object.
+        editor_name : string
+            The name of the relevant editor trait.
+        is_list : bool, optional
+            If true, synchronization for item events will be set up in
+            addition to the synchronization for the object itself.
+            The default is False.
+        """
+
+        def editor_trait_modified(new):
+            if key not in self._no_trait_update:
+                with self.no_trait_update(key), self.raise_to_debug():
+                    xsetattr(user_object, xuser_name, new)
+
+        self.on_trait_change(editor_trait_modified, editor_name)
+
+        if self._user_from is None:
```

Comment:
I'll refrain from commenting on this, since it's unchanged by the PR ...

D'oh!

Reply:
But it does look as though all the `is None` and `is not None` checks on `self._user_to` and `self._user_from` should be unnecessary, now that we have the trait defaults.

Reply:
This was `HasPrivateTraits` nonsense, but now that the traits are properly defined, it's not needed.